File: algorithms/388. Longest Absolute File Path/solution.py
class Solution(object):
    def lengthLongestPath(self, input):
        """
        :type input: str
        :rtype: int
        """
        # @StefanPochmann
        maxlen = 0
        pathlen = {0: 0}
        for file in input.splitlines():
            name = file.lstrip('\t')
            depth = len(file) - len(name)
            if '.' in name:
                maxlen = max(maxlen, pathlen[depth] + len(name))
            else:
                pathlen[depth + 1] = pathlen[depth] + len(name) + 1
        return maxlen

# 25 / 25 test cases passed.
# Status: Accepted
# Runtime: 58 ms
# beats 15.59 %

File: algorithms/388. Longest Absolute File Path/solution.cc
class Solution {
public:
    int lengthLongestPath(string input) {
        stringstream ss(input);
        string token;
        int maxlen = 0;
        map<int, int> pathlen({{0, 0}});
        while ( getline(ss, token) ) {
            auto pos = token.find_last_of('\t');
            string name = pos == string::npos ? token : token.substr(pos + 1);
            int depth = token.size() - name.size();
            if ( name.find('.') != string::npos ) {
                maxlen = max(maxlen, pathlen[depth] + int(name.size()));
            } else {
                pathlen[depth + 1] = pathlen[depth] + name.size() + 1;
            }
        }
        return maxlen;
    }
};
